cpp/src/community/edge_triangle_count_impl.cuh

```diff
@@ -0,0 +1,168 @@
+#include <cugraph/graph_view.hpp>
+#include <cugraph/partition_manager.hpp>
+#include <cugraph/utilities/error.hpp>
+#include <cugraph/utilities/mask_utils.cuh>
+
+#include <raft/core/handle.hpp>
+#include <raft/util/cudart_utils.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/adjacent_difference.h>
+#include <thrust/binary_search.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+
+#include <detail/graph_partition_utils.cuh>
+#include <prims/transform_reduce_dst_nbr_intersection_of_e_endpoints_by_v.cuh>
+
+#include <optional>
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+struct update_edges_p_r_q_r_num_triangles {
+  size_t num_edges{};  // rename to num_edges
+  const edge_t edge_first_or_second{};
+  raft::device_span<size_t const> intersection_offsets{};
+  raft::device_span<vertex_t const> intersection_indices{};
+  raft::device_span<edge_t> num_triangles{};
+
+  EdgeIterator edge_first{};
+
+  __device__ void operator()(size_t i) const
+  {
+    auto itr = thrust::upper_bound(
+      thrust::seq, intersection_offsets.begin() + 1, intersection_offsets.end(), i);
+    auto idx = thrust::distance(intersection_offsets.begin() + 1, itr);
+    if (edge_first_or_second == 0) {
+      auto p_r_pair =
+        thrust::make_tuple(thrust::get<0>(*(edge_first + idx)), intersection_indices[i]);
+
+      // Find its position in 'edges'
+      auto itr_p_r_p_q =
+        thrust::lower_bound(thrust::seq,
+                            edge_first,
+                            edge_first + num_edges,  // pass the number of vertex pairs
+                            p_r_pair);
+
+      assert(*itr_p_r_p_q == p_r_pair);
+      idx = thrust::distance(edge_first, itr_p_r_p_q);
+    } else {
+      auto p_r_pair =
+        thrust::make_tuple(thrust::get<1>(*(edge_first + idx)), intersection_indices[i]);
+
+      // Find its position in 'edges'
+      auto itr_p_r_p_q =
+        thrust::lower_bound(thrust::seq,
+                            edge_first,
+                            edge_first + num_edges,  // pass the number of vertex pairs
+                            p_r_pair);
+      assert(*itr_p_r_p_q == p_r_pair);
+      idx = thrust::distance(edge_first, itr_p_r_p_q);
+    }
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_add(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t, typename edge_t, bool store_transposed, bool multi_gpu>
+std::enable_if_t<!multi_gpu, rmm::device_uvector<edge_t>> edge_triangle_count_impl(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  raft::device_span<vertex_t> edgelist_srcs,
+  raft::device_span<vertex_t> edgelist_dsts)
+{
```

Comment:
Add a FIXME statement saying we need to use chunking to limit peak memory requirement. We can address in a separate PR focusing on edge triangle count.